### PR TITLE
Add Vulkan support

### PR DESCRIPTION
**Pull request by mbouron-gpsw**

### CODE REVIEW

File: configure.py [hunk 1]
```diff
@@ -64,11 +65,17 @@
...
+        dst_file='glslang-@VERSION@.tar.gz',
+        url='https://github.com/KhronosGroup/glslang/archive/refs/tags/@VERSION@.tar.gz',
+        sha256='9e5fbe5b844d203da5e61bcd84eda76326e0ff5dc696cb862147bbe01d2febb0',
+    ),
 )
 
 
 def _get_external_deps(args):
-    deps = ['sxplayer']
+    deps = ['sxplayer', 'glslang']
```
Comment (cboesch-gpsw) [inline configure.py:78]: shouldn't that be in the windows scope below?

File: configure.py [hunk 2]
```diff
@@ -97,6 +115,26 @@ def _file_chk(path, chksum_hexdigest):
...
+        for file in files:
+            os.chmod(os.path.join(root, file), stat.S_IRUSR | stat.S_IWUSR)
+        for directory in dirs:
+            os.chmod(os.path.join(root, directory), stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)
+
+
+def _rmtree(path, ignore_errors=False, onerror=None):
+    """
+    shutil.rmtree wrapper that is resilient to permission issues when
+    encoutering read-only files or directories lacking the executable
```
Comment (cboesch-gpsw) [inline configure.py:129]: encountering

File: configure.py [hunk 3]
```diff
@@ -368,6 +426,37 @@ def _tests(cfg):
     return ['$(MESON) ' + _cmd_join('test', '-C', op.join('builddir', 'tests'))]
 
 
+@_block('glslang-setup')
+def _glslang_setup(cfg):
+    srcdir = cfg.externals['glslang']
+    builddir = op.join('builddir', 'glslang')
+
+    # Pull internal dependencies
+    run(['python', 'update_glslang_sources.py'], cwd=srcdir, check=True)
```
Comment (cboesch-gpsw) [inline configure.py:435]: If this requires `git` to be installed on Windows, we need to add it to the documentation

File: configure.py [hunk 4]
```diff
@@ -368,6 +426,37 @@ def _tests(cfg):
...
+
+    # Shared build is not supported on Windows
+    build_shared = 'OFF' if _SYSTEM == 'Windows' else 'ON'
+    opts = [
+        '-DCMAKE_BUILD_TYPE=Release',
```
Comment (cboesch-gpsw) [inline configure.py:440]: this conflicts with the settings set in `CMAKE_SETUP`

File: configure.py [hunk 5]
```diff
@@ -368,6 +426,37 @@ def _tests(cfg):
...
+        f'-DBUILD_SHARED_LIBS={build_shared}',
+        '-DENABLE_OPT=ON',
+        '-DBUILD_EXTERNAL=ON',
+        '-DENABLE_SPVREMAPPER=OFF',
+        '-DENABLE_GLSLANG_BINARIES=OFF',
+    ]
+    return [
+        f'$(CMAKE_SETUP) -S {srcdir} -B {builddir} ' + shlex.join(opts),
+    ]
```
Comment (cboesch-gpsw) [inline configure.py:449]: nit: one line

File: configure.py [hunk 6]
```diff
@@ -368,6 +426,37 @@ def _tests(cfg):
...
+
+
+@_block('glslang-install', [_glslang_setup])
+def _glslang_install(cfg):
+    builddir = op.join('builddir', 'glslang')
+    return [
+        f'cmake --build {builddir} --target install',
+    ]
```
Comment (cboesch-gpsw) [inline configure.py:457]: ditto 1 line

File: configure.py [hunk 7]
```diff
@@ -569,6 +667,8 @@ def _run():
         _all, _tests, _clean,
         _nodegl_updatedoc, _nodegl_updatespecs, _nodegl_updateglwrappers,
     ]
+    if _SYSTEM == 'Windows':
+        blocks += [_glslang_install]
```
Comment (cboesch-gpsw) [inline configure.py:671]: This needs to be a dependency of `nodegl-setup` not a generic rule; how is it triggered today?

Is the re-entrance a fast no-op when already installed?

File: libnodegl/api.c [hunk 8]
```diff
@@ -57,7 +57,11 @@
 #if defined(TARGET_IPHONE) || defined(TARGET_ANDROID)
 # define DEFAULT_BACKEND NGL_BACKEND_OPENGLES
 #else
-# define DEFAULT_BACKEND NGL_BACKEND_OPENGL
+# if defined(BACKEND_VK)
+#  define DEFAULT_BACKEND NGL_BACKEND_VULKAN
```
Comment (cboesch-gpsw) [inline libnodegl/api.c:61]: Do we want vulkan by default right now?

File: pynodegl-utils/pynodegl_utils/config.py [hunk 9]
```diff
@@ -69,7 +70,7 @@ def __init__(self, module_pkgname):
             'framerate': (60, 1),
             'log_level': 'info',
             'clear_color': (0.0, 0.0, 0.0, 1.0),
-            'backend': 'opengl',
+            'backend': 'vulkan',
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/config.py:73]: ditto vulkan by default (what does it work properly when vulkan is not available?)

File: libnodegl/meson.build [hunk 10]
```diff
@@ -362,22 +363,172 @@ gbackends_cfg = {
...
+    # from the system library directories on macOS and Windows, and the
+    # directories provided by the user through the dirs argument. For example,
+    # on macOS (Intel), the system library directories contains /usr/lib and
+    # /usr/local/lib whereas the compiler search directories does not. This is
+    # not an issue on Linux as gcc or clang search directories contains
+    # /usr/lib and /usr/local/lib.
+    # To workaround this limitation, we first try to find the specified
+    # library without specifying the dirs argument. Then, if the library is not
+    # found and not optional we try to find it using the additional search
+    # paths provided by the user (through extra_library_dirs).
```
Comment (cboesch-gpsw) [inline libnodegl/meson.build:523]: Shouldn't we open an issue/discussion on meson tracker and link it in the code?

File: libnodegl/meson_options.txt [hunk 11]
```diff
@@ -36,8 +37,14 @@ option('tests', type: 'boolean', value: true)
...
-option('debug_opts', type: 'array', choices: ['gl', 'mem', 'scene', 'gpu_capture'], value: [],
+option('debug_opts', type: 'array', choices: ['gl', 'vk', 'mem', 'scene', 'gpu_capture'], value: [],
        description: 'debugging options for developers')
 
 option('renderdoc_dir', type: 'string',
        description: 'Renderdoc directory')
+
+option('extra_library_dirs', type: 'array',
+       description: 'list of directories to look from for external libraries', value: [])
+
```
Comment (cboesch-gpsw) [inline libnodegl/meson_options.txt:48]: please pack those together by removing this line break

Also "from for" is weird, maybe "into for"?

File: CHANGELOG.md [hunk 12]
```diff
@@ -11,6 +11,7 @@ Versioning](https://semver.org/spec/v2.0.0.html) for `libnodegl`.
 ### Added
 - TextureView node, useful for targeting a specific layer of a texture in RTT
   scenarios
+- Initial Vulkan support
```
Comment (cboesch-gpsw) [inline CHANGELOG.md:14]: Maybe "- Initial Vulkan support, including Android video acceleration"

Comment (mbouron-gpsw) [reply CHANGELOG.md:14]: The Android video interop is not ready yet, I will be part of a dedicated PR soon after the initial merge.

File: configure.py (hunk 5 above)
Comment (mbouron-gpsw) [reply configure.py]: Fixed.

File: configure.py (hunk 6 above)
Comment (mbouron-gpsw) [reply configure.py]: Fixed.

File: libnodegl/api.c (hunk 8 above)
Comment (mbouron-gpsw) [reply libnodegl/api.c]: Nope, fixed.

File: pynodegl-utils/pynodegl_utils/config.py (hunk 9 above)
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/config.py]: Fixed.

File: libnodegl/meson_options.txt (hunk 11 above)
Comment (mbouron-gpsw) [reply libnodegl/meson_options.txt]: > please pack those together by removing this line break

Done.

File: configure.py (hunk 1 above)
Comment (mbouron-gpsw) [reply configure.py]: Fixed.

File: configure.py (hunk 2 above)
Comment (mbouron-gpsw) [reply configure.py]: Fixed.

File: configure.py (hunk 4 above)
Comment (mbouron-gpsw) [reply configure.py]: Fixed.

File: pynodegl-utils/pynodegl_utils/misc.py [hunk 13]
```diff
@@ -146,7 +146,7 @@ class SceneCfg:
         'aspect_ratio': (16, 9),
         'duration': 30.0,
         'framerate': (60, 1),
-        'backend': 'opengl',
+        'backend': 'vulkan',
```
Comment (cboesch-gpsw) [inline pynodegl-utils/pynodegl_utils/misc.py:149]: another default that you may want to preserve as opengl

File: libnodegl/node_media.c [hunk 14]
```diff
@@ -207,7 +207,6 @@ static int media_init(struct ngl_node *node)
     struct android_ctx *android_ctx = &ctx->android_ctx;
     const struct ngl_config *config = &ctx->config;
 
-    if (config->backend == NGL_BACKEND_OPENGLES) {
         void *android_surface = NULL;
         if (android_ctx->has_native_imagereader_api) {
```
Comment (cboesch-gpsw) [inline libnodegl/node_media.c:211]: no reindent?

Comment (mbouron-gpsw) [reply libnodegl/node_media.c]: Done in a separate commit.

File: pynodegl-utils/pynodegl_utils/misc.py (hunk 13 above)
Comment (mbouron-gpsw) [reply pynodegl-utils/pynodegl_utils/misc.py]: Done.

File: configure.py (hunk 7 above)
Comment (mbouron-gpsw) [reply configure.py]: I removed the build of glslang from the venv, it relies now on vcpkg to install it.

File: configure.py (hunk 3 above)
Comment (mbouron-gpsw) [reply configure.py]: glslang is no more installed by the venv.

File: libnodegl/meson.build (hunk 10 above)
Comment (mbouron-gpsw) [reply libnodegl/meson.build:523]: I guess that what should at least start a discussion about this.

**Review by cboesch-gpsw — APPROVED** [review]

**Review by wmacdonald-gpsw — APPROVED** [review]

**Pull request merged**